packages/autocomplete-js/src/__tests__/api.test.ts

```diff
@@ -445,6 +447,73 @@ describe('api', () => {
         expect(mockCreateElement2).toHaveBeenCalled();
       });
     });
+
+    test('preserves all user `components` when not updated', async () => {
+      const container = document.createElement('div');
+      const panelContainer = document.createElement('div');
+      document.body.appendChild(container);
+      document.body.appendChild(panelContainer);
+
+      const CustomFragment = (props: any) => props.children;
+      const mockCreateElement1 = jest.fn(preactCreateElement);
+      const mockCreateElement2 = jest.fn(preactCreateElement);
+      const mockRender = jest.fn().mockImplementation(preactRender);
+      const CustomHighlight = jest.fn((props: { hit: { label: string } }) =>
+        mockCreateElement1(CustomFragment, null, props.hit.label)
+      );
+      const MyComponent = (props: any) => props.children;
+
+      const { update } = autocomplete<{ label: string }>({
+        container,
+        panelContainer,
+        getSources() {
+          return [
+            {
+              sourceId: 'testSource',
+              getItems({ query }) {
+                return [{ label: query }];
+              },
+              templates: {
+                item({ item, components, html }) {
+                  return html`<div>
+                    ${components.Highlight({ hit: item, attribute: 'label' })}
+                    ${components.MyComponent({ children: item.label })}
+                  </div>`;
+                },
+              },
+            },
+          ];
+        },
+        components: { Highlight: CustomHighlight, MyComponent },
+        renderer: {
+          Fragment: CustomFragment,
+          render: mockRender,
+          createElement: mockCreateElement1,
+        },
+      });
+
+      update({
+        renderer: {
+          Fragment: CustomFragment,
+          render: mockRender,
+          createElement: mockCreateElement2,
+        },
+      });
+
+      const input = container.querySelector<HTMLInputElement>('.aa-Input');
+
+      fireEvent.input(input, { target: { value: 'iphone' } });
+
+      await waitFor(() => {
+        expect(
+          panelContainer.querySelector<HTMLElement>('.aa-Panel')
+        ).toHaveTextContent('iphone');
+        // The custom `Highlight` component wasn't updated, so the previous
+        // `createElement` implementation is still being called.
+        expect(mockCreateElement1).toHaveBeenCalled();
+        expect(mockCreateElement2).toHaveBeenCalled();
```

Comment:
Can we add a clarification in the test like reset the mock count before `input()` and call `toHaveBeenCalledTimes(1)`? It would make the test more resilient if we're to use `createElement` in the search box or anywhere else.

Reply:
Good idea, addressed in ee415fdb.